Port SVD to ATen, enable batching for matrix inputs

vishwakftw opened the pull request:
Changelog:
- Port SVD TH implementation to ATen/native/BatchLinearAlgebra.cpp
- Port SVD THC implementation to ATen/native/cuda/BatchLinearAlgebra.cu
- Allow batches of matrices as arguments to `torch.svd`
- Remove existing implementations in TH and THC
- Update doc string
- Update derivatives to support batching
- Modify nuclear norm implementation to use at::svd instead of _batch_svd
- Remove _batch_svd as it is redundant

Test Plan:
- Add new test suite for SVD in test_torch.py with port to test_cuda.py
- Add tests in common_methods_invocations.py for derivative testing

vishwakftw committed:
Port TH SVD to ATen, enable batching

vishwakftw commented:
Test failures are expected. I removed the TH bindings.

vishwakftw committed:
Port THC SVD to ATen, enable batching

vishwakftw committed:
Rebase commit

vishwakftw committed:
Add bindings to svd

vishwakftw commented:
I’ll investigate the test failures, they are related to this PR.

vishwakftw committed:
Fix test_lapack_empty

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-svd

vishwakftw committed:
Fix CUDA failures, add tests and fix after local failures. Remove _batch_svd function and use at::svd instead

vishwakftw committed:
Update derivative

vishwakftw committed:
Update doc strings

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-svd

vishwakftw committed:
Fix strides mismatch between CPU and CUDA

vishwakftw commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

vishwakftw commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

vishwakftw committed:
Don't expose _svd_helper to python API

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-svd

gchanan commented:
@nairbv can you review this, please?

vishwakftw committed:
Don't use pinned_memory

nairbv commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -247,4 +251,55 @@ static inline std::tuple<std::vector<int64_t>,
+  if (!input.is_cuda()) {
+    U_empty = at::empty_strided(sizes, strides, input.options());
+  } else {
+    U_empty = at::empty_strided(sizes, strides, input.options().device(at::kCPU));
```
why does these need to be on CPU tensors? Do they get moved back to cuda later?

vishwakftw replied:
Yes, they are moved back in _svd_helper_cuda.

The MAGMA routine accepts matrices on CPU only, since it is a hybrid algorithm.

nairbv replied:
ah OK, that makes sense. Maybe we can add that explanation in a code comment somewhere -- it's a bit confusing, at least to me, to see the `to(kCPU)` in a cuda op.

vishwakftw replied:
Yep, I will add a comment.

nairbv commented on torch/_torch_docs.py:
```diff
@@ -4888,47 +4882,44 @@ def merge_dicts(*dicts):
-                          [5.45, -0.27,  4.85,  0.74, 10.00, -6.02],
-                          [3.16,  7.98,  3.01,  5.80,  4.27, -5.31]]).t()
-
+    >>> a = torch.randn(5, 3)
```
I think it's helpful having an actual tensor in the doc that can be copy/pasted to reproduce the same documented output

vishwakftw replied:
Sure, I will make that chane.

nairbv commented on torch/_torch_docs.py:
```diff
@@ -4857,29 +4857,23 @@ def merge_dicts(*dicts):
 
 .. note:: Irrespective of the original strides, the returned matrix `U`
-          will be transposed, i.e. with strides `(1, n)` instead of `(n, 1)`.
+          will be transposed, i.e. with strides :code:`U.contiguous().transpose(-2, -1).strides()`
```
I think you want `stride` not `strides` here

vishwakftw replied:
Sorry, I’ll fix it. Thank you for pointing it out.

vishwakftw committed:
Code Review - I

- Add a comment about why empty tensors are created on the CPU while the input is a CUDA tensor in _create_U_S_VT
- Print matrix in SVD doc example
- strides() --> stride() in docs

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-svd

nairbv commented:
@pytorchbot rebase this please

nairbv commented:
looks like failing tests were due to whatever this fixed:
https://github.com/pytorch/pytorch/commit/b5fa9a340a0d174131ad0a452c395860d571b5b0

vishwakftw commented:
Yes, I’ll manually rebase.

vishwakftw committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into batch-svd

nairbv commented:
LGTM
